mobile_verifier/src/data_session.rs

```diff
@@ -0,0 +1,243 @@
+    pub async fn save(self, db: &mut Transaction<'_, Postgres>) -> Result<(), DataSessionError> {
+        sqlx::query(
+            r#"
+            INSERT INTO hotspot_data_transfer_sessions (pub_key, payer, upload_bytes, download_bytes, num_dcs, reward_timestamp)
+            VALUES ($1, $2, $3, $4, $5, $6)
+            "#,
+        )
+        .bind(self.pub_key)
+        .bind(self.payer)
+        .bind(self.upload_bytes)
+        .bind(self.download_bytes)
+        .bind(self.num_dcs)
+        .bind(self.reward_timestamp)
+        .execute(&mut *db)
+        .await?;
+        Ok(())
+    }
+    fn from_valid_data_session(
+        v: ValidDataTransferSession,
+        reward_timestamp: DateTime<Utc>,
+    ) -> HotspotDataSession {
+        Self {
+            pub_key: v.pub_key,
+            payer: v.payer,
+            upload_bytes: v.upload_bytes as i64,
+            download_bytes: v.download_bytes as i64,
+            num_dcs: v.num_dcs as i64,
+            reward_timestamp,
+        }
+    }
+}
+
+pub async fn aggregate_hotspot_data_sessions_to_dc<'a>(
+    exec: impl sqlx::PgExecutor<'a> + Copy + 'a,
+    epoch: &'a Range<DateTime<Utc>>,
+) -> Result<HotspotMap, sqlx::Error> {
+    let stream = sqlx::query_as::<_, HotspotDataSession>(
+        r#"
+        SELECT *
+        FROM hotspot_data_transfer_sessions
+        WHERE reward_timestamp >= $1 and reward_timestamp < $2
+        "#,
+    )
+    .bind(epoch.start)
+    .bind(epoch.end)
+    .fetch(exec);
+    data_sessions_to_dc(stream).await
+}
+
+pub async fn data_sessions_to_dc<'a>(
+    stream: impl Stream<Item = Result<HotspotDataSession, sqlx::Error>>,
+) -> Result<HotspotMap, sqlx::Error> {
+    tokio::pin!(stream);
+    let mut map = HotspotMap::new();
+    while let Some(session) = stream.try_next().await? {
+        *map.entry(session.pub_key).or_default() += session.num_dcs as u64
+    }
+    Ok(map)
+}
+
+pub async fn aggregate_subscriber_data_sessions_to_bytes<'a>(
+    exec: impl sqlx::PgExecutor<'a> + Copy + 'a,
+    epoch: &Range<DateTime<Utc>>,
+) -> Result<SubscriberMap, sqlx::Error> {
+    let stream = sqlx::query_as::<_, SubscriberDataSession>(
+        r#"
+            SELECT *
+            FROM subscriber_data_transfer_sessions
+            WHERE download_bytes > 0 OR upload_bytes > 0
+            AND reward_timestamp >= $1 AND reward_timestamp < $2
+            "#,
+    )
+    .bind(epoch.start)
+    .bind(epoch.end)
+    .fetch(exec);
+    data_sessions_to_bytes(stream).await
+}
+
+pub async fn data_sessions_to_bytes<'a>(
+    stream: impl Stream<Item = Result<SubscriberDataSession, sqlx::Error>>,
+) -> Result<SubscriberMap, sqlx::Error> {
+    tokio::pin!(stream);
+    let mut map = SubscriberMap::new();
+    while let Some(session) = stream.try_next().await? {
+        *map.entry(session.subscriber_id).or_default() +=
+            (session.download_bytes + session.upload_bytes) as u64
+    }
+    Ok(map)
+}
+
+pub async fn clear_subscriber_data_sessions(
```

Comment:
i don't see us clearing hotspot data sessions? do we need to wipe those from the database once the rewards for the covered period have been generated as well?